### PR TITLE
[patch][SDK-1801] Collect, Persist, and Forward 'gclid' value on downstream events

### PR DESCRIPTION
**Pull request by nsingh-branch**
## Reference
SDK-1801 -- Collect, Persist, and Forward 'gclid' value on downstream events

## Description
Updated the way we collect and use gclid and other url query parameters by creating a new class, ReferringUrlUtility. 

### Overview
1. A user opens the app from a link containing url query parameters (eg. [timber.app.link/h34jfkd4?gclid=123abc](url))
2. Upon app open, `parseReferringURL()` is called from `ServerRequest.java` and the url is scanned for any supported URL query parameters. If found, the parameter is saved along with its timestamp and default validity window as a `BranchUrlQueryParameter`.
3. Now whenever a post request is made, `getURLQueryParamsForRequest()` is called from `ServerRequest.java` which adds the respective query parameters depending on the type of request.
   a. Each supported query parameter has it's own method (eg. `addGclidValueFor()`) which adds the required fields to a JSONObject.  

**Example JSONObject added to requests:**
```
{
"gclid": 123abc,
"is_deeplink_gclid: true
}
```

## Testing Instructions
- Add gclid as a query parameter to a referring URL, eg. https://bnctestbed.app.link?gclid=12345
- Look for gclid values in the payload for v1/opens and v2/events. Also is_deeplink_gclid in v1/open

## Risk Assessment [`LOW`]

- [x] I, the PR creator, have tested — integration, unit, or otherwise — this code.

## Reviewer Checklist (To be checked off by the reviewer only)

- [ ] JIRA Ticket is referenced in PR title.
- Correctness & Style
    - [ ] Conforms to [AOSP Style Guides](https://source.android.com/setup/contribute/code-style)
    - [ ] Mission critical pieces are documented in code and out of code as needed.
- [ ] Unit Tests reviewed and test issue sufficiently.
- [ ] Functionality was reviewed in QA independently by another engineer on the team.

cc @BranchMetrics/saas-sdk-devs for visibility.

### CODE REVIEW

**Comment by gdeluna-branch** [comment]
In your description, can you summarize at a high level the algorithm? With some example inputs and outputs.

File: Branch-SDK-TestBed/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -66,7 +66,7 @@
             <meta-data android:name="io.branch.sdk.auto_link_request_code" android:value="@integer/AutoDeeplinkRequestCode" />
         </activity>
 
-        <meta-data android:name="io.branch.sdk.TestMode" android:value="true" /> <!-- Set to true to use Branch_Test_Key -->
+        <meta-data android:name="io.branch.sdk.TestMode" android:value="false" /> <!-- Set to true to use Branch_Test_Key -->
```
Comment (gdeluna-branch) [inline Branch-SDK-TestBed/src/main/AndroidManifest.xml:69]: Just curious if there's a reason, it's fine to switch it.

File: Branch-SDK/src/main/java/io/branch/referral/DeviceInfo.java [hunk 2]
```diff
@@ -212,13 +215,11 @@ void updateRequestWithV2Params(ServerRequest serverRequest, PrefHelper prefHelpe
...
     void updateRequestWithParamsAllEvents(ServerRequest serverRequest, PrefHelper prefHelper, JSONObject requestObj){
         try {
-            // For install events, referrer GCLID is already contained in `install_referrer_extras`
-            // Otherwise, for all other v1 and v2 events, add referrer_gclid to top level
-            if (!(serverRequest instanceof ServerRequestRegisterInstall)) {
-                String gclid = prefHelper.getReferrerGclid();
-                if (gclid != null && !gclid.equals(NO_STRING_VALUE)) {
-                    requestObj.put(Defines.Jsonkey.ReferrerGclid.getKey(), gclid);
-                }
+            Log.d("Testing", "updateRequestWithParamsAllEvents");
```
Comment (gdeluna-branch) [inline Branch-SDK/src/main/java/io/branch/referral/DeviceInfo.java:218]: Use prefhelper.debug wrapper

Comment (gdeluna-branch) [reply Branch-SDK/src/main/java/io/branch/referral/DeviceInfo.java:218]: or remove if unneeded

File: Branch-SDK/src/main/java/io/branch/referral/PrefHelper.java [hunk 3]
```diff
@@ -706,6 +706,33 @@ public String getAppStoreSource(){
         return getString(KEY_APP_STORE_SOURCE);
     }
 
+    /**
+     * Sets the referring URL query parameters.
+     * @param referringUrlQueryParameters
+     */
+    public void setReferringUrlQueryParameters(JSONObject referringUrlQueryParameters) {
+        setString("bnc_referringUrlQueryParameters", String.valueOf(referringUrlQueryParameters));
```
Comment (gdeluna-branch) [inline Branch-SDK/src/main/java/io/branch/referral/PrefHelper.java:714]: Make `bnc_referringUrlQueryParameters` a key

File: Branch-SDK/src/main/java/io/branch/referral/PrefHelper.java [hunk 4]
```diff
@@ -706,6 +706,33 @@ public String getAppStoreSource(){
...
+    }
+
+    /**
+     * Returns the referring URL query parameters.
+     * @return
+     */
+    public JSONObject getReferringURLQueryParameters()  {
+
+        String string = getString("bnc_referringUrlQueryParameters");
```
Comment (gdeluna-branch) [inline Branch-SDK/src/main/java/io/branch/referral/PrefHelper.java:723]: And use said constant where needed

File: Branch-SDK/src/main/java/io/branch/referral/ReferringUrlUtility.kt [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+                // If there is no validity window, set to default.
+                if (param.validityWindow == 0L) {
+                    param.validityWindow = defaultValidityWindowForParam(paramName) as Long
+                }
+
+                urlQueryParameters[paramName] = param
+            }
+        }
+
+        prefHelper.setReferringUrlQueryParameters(serializeToJson(urlQueryParameters))
```
Comment (gdeluna-branch) [inline Branch-SDK/src/main/java/io/branch/referral/ReferringUrlUtility.kt:42]: Do you need to check for a previously existing gclid on the device?

File: Branch-SDK/src/main/java/io/branch/referral/ReferringUrlUtility.kt [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+                    prefHelper.setReferringUrlQueryParameters(serializeToJson(urlQueryParameters))
+                }
+            }
+        }
+
+        return returnParams
+    }
+
+    @VisibleForTesting
+    internal fun isSupportedQueryParameter(paramName: String): Boolean {
```
Comment (gdeluna-branch) [inline Branch-SDK/src/main/java/io/branch/referral/ReferringUrlUtility.kt:82]: Some functions are marked `internal`, another `private`, and some no modifier. No modifier in Kotlin is `public`, as opposed to Java which is `package-private`.

https://medium.com/@HugoMatilla/kotlin-basics-visibility-modifiers-public-internal-protected-and-private-c3bf972aee11

Regarding `@VisibleForTesting`: https://timkranen.medium.com/using-visiblefortesting-to-test-your-private-functions-c9e511fd1c2a

The tl;dr is (as i interpret it) be mindful of testing private functions since the public methods that use the results of them should be tested anyway and give you the same value (the outputs of your private functions). The tests are fine to keep but just something to keep in mind when thinking of how to write unit test.

File: Branch-SDK/src/main/java/io/branch/referral/ReferringUrlUtility.kt [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+            param.name?.let { paramName ->
+                result[paramName] = param
+            }
+        }
+
+        return result
+    }
+
+    @VisibleForTesting
+    internal fun checkForAndMigrateOldGclid() {
```
Comment (gdeluna-branch) [inline Branch-SDK/src/main/java/io/branch/referral/ReferringUrlUtility.kt:157]: Can you method-comment what the logic is here?

File: Branch-SDK/src/main/java/io/branch/referral/ReferringUrlUtility.kt [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+                    value = existingGclidValue,
+                    timestamp = Date(),
+                    validityWindow = existingGclidValidityWindow,
+                    isDeepLink = false
+                )
+
+                urlQueryParameters[Defines.Jsonkey.Gclid.key] = gclid
+
+                prefHelper.setReferringUrlQueryParameters(serializeToJson(urlQueryParameters))
+                prefHelper.referrerGclid = null
```
Comment (gdeluna-branch) [inline Branch-SDK/src/main/java/io/branch/referral/ReferringUrlUtility.kt:176]: I think there was a whole `JSONObject` that encompassed both sub JSONObjects, would it make sense to null that out instead?

File: Branch-SDK/src/main/java/io/branch/referral/ServerRequest.java [hunk 9]
```diff
@@ -675,7 +678,22 @@ public boolean isWaitingOnProcessToFinish() {
      * Called on UI thread just before executing a request. Do any final updates to the request here
      */
     public void onPreExecute() {
-    
+        Log.e("Testing", "onPreExecute" + this + "" + (this instanceof ServerRequestRegisterOpen || this instanceof BranchEvent.ServerRequestLogEvent));
```
Comment (gdeluna-branch) [inline Branch-SDK/src/main/java/io/branch/referral/ServerRequest.java:681]: Is this test log needed?

File: Branch-SDK/src/main/java/io/branch/referral/util/BranchEvent.java [hunk 10]
```diff
@@ -256,7 +256,7 @@ public boolean logEvent(Context context) {
         return isReqQueued;
     }
 
-    private class ServerRequestLogEvent extends ServerRequest {
+    public class ServerRequestLogEvent extends ServerRequest {
```
Comment (gdeluna-branch) [inline Branch-SDK/src/main/java/io/branch/referral/util/BranchEvent.java:259]: Can you extract this out to its own file? I'd personally like to get rid of nested classes for easier project readability.

**Review by gdeluna-branch — CHANGES_REQUESTED** [review]
mostly cleanup related

**Comment by gdeluna-branch** [comment]
Pull latest/validate tests and I'll approve

**Review by gdeluna-branch — APPROVED** [review]
